### DISCUSSION TITLE
tutorials/intro-dgs/11-mutations

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#132] [course feedback]
# tutorials/intro-dgs/11-mutations

https://www.apollographql.com/tutorials/intro-dgs/11-mutations

**Comment by punjprakash** [comment]
We need to give return type when snapshot is null.
@DgsMutation
    public AddItemsToPlaylistPayload addItemsToPlaylist(@InputArgument AddItemsToPlaylistInput input) {
        String playlistId = input.getPlaylistId();
        List<String> uris = input.getUris();

        Snapshot snapshot = spotifyClient.addItemsToPlaylist(playlistId, String.join(",", uris));

        AddItemsToPlaylistPayload payload = new AddItemsToPlaylistPayload();

        if (snapshot != null) {
            String snapshotId = snapshot.id();
            if (Objects.equals(snapshotId, playlistId)) {
                Playlist playlist = new Playlist();
                playlist.setId(playlistId);
                payload.setCode(200);
                payload.setMessage("success");
                payload.setSuccess(true);
                payload.setPlaylist(playlist);
                return payload;
            }
        }
        payload.setCode(500);
        payload.setMessage("could not update playlist");
        payload.setSuccess(false);
        payload.setPlaylist(null);
        return payload;
    };

**Reply by lizhennessy** [reply]
Hi there @punjprakash ! I'm not sure I understand - in this example, we set the various playlist, status code, and message details of the `AddItemsToPlaylistPayload` instance when snapshot returns with data. In the event that it returns as _null_, we set the code as `500`, the failure message, and the playlist as null. Are you looking for a more robust return type than an `AddItemsToPlaylistPayload` instance for this method?

**Reply by punjprakash** [reply]
Sorry I was not clear with my question. In the method "addItemsToPlaylist"  you have return in block "if(snapshot!=null){}" but we need to have return outside of the  "if(snapshot!=null){}" as well in case snapshot is null. Right now, IDE complaining about missing return. Below is your code:

```
@DgsMutation
public AddItemsToPlaylistPayload addItemsToPlaylist(@InputArgument AddItemsToPlaylistInput input) {
  String playlistId = input.getPlaylistId();
  List<String> uris = input.getUris();
  
  Snapshot snapshot = spotifyClient.addItemsToPlaylist(playlistId, String.join(",", uris));
  
  AddItemsToPlaylistPayload payload = new AddItemsToPlaylistPayload();
  
  if (snapshot !== null) {
    String snapshotId = snapshot.id();
    if (Objects.equals(snapshotId, playlistId)) {
      Playlist playlist = new Playlist();
      playlist.setId(playlistId);
      payload.setCode(200);
      payload.setMessage("success");
      payload.setSuccess(true);
      payload.setPlaylist(playlist);
      return payload;
    }
    payload.setCode(500);
    payload.setMessage("could not update playlist");
    payload.setSuccess(false);
    payload.setPlaylist(null);
    return payload;
  }
};
```

**Reply by punjprakash** [reply]
I can see the correct code in the repo https://github.com/apollographql-education/intro-dgs/blob/final/src/main/java/com/example/soundtracks/datafetchers/PlaylistDataFetcher.java

```


    @DgsMutation
    public AddItemsToPlaylistPayload addItemsToPlaylist(@InputArgument AddItemsToPlaylistInput input) {
        String playlistId = input.getPlaylistId();
        List<String> uris = input.getUris();

        Snapshot snapshot = spotifyClient.addItemsToPlaylist(playlistId, String.join(",", uris));

        AddItemsToPlaylistPayload payload = new AddItemsToPlaylistPayload();

        if (snapshot != null) {
            String snapshotId = snapshot.id();
            if (Objects.equals(snapshotId, playlistId)) {
                Playlist playlist = new Playlist();
                playlist.setId(playlistId);

                payload.setCode(200);
                payload.setMessage("success");
                payload.setSuccess(true);
                payload.setPlaylist(playlist);

                return payload;
            }
        }
        payload.setCode(500);
        payload.setMessage("could not update playlist");
        payload.setSuccess(false);
        payload.setPlaylist(null);

        return payload;
    }
```

**Reply by lizhennessy** [reply]
Ohhh, I see thank you for pointing this out @punjprakash ! I'll take a look and get a fix out.
